futures-select-macro/src/lib.rs

```diff
@@ -168,13 +168,25 @@ pub fn select(input: TokenStream) -> TokenStream {
         .zip(variant_names.iter())
         .map(|(expr, variant_name)| {
             match expr {
-                // Don't bind futures that are already a path.
-                // This prevents creating redundant stack space
-                // for them.
-                syn::Expr::Path(path) => path,
+                syn::Expr::Path(path) => {
+                    // Don't bind futures that are already a path.
+                    // This prevents creating redundant stack space
+                    // for them.
+                    path
+                },
                 _ => {
+                    // Bind and pin the resulting Future on the stack. This is
+                    // necessary to support direct select! calls on !Unpin
+                    // Futures.
+                    // Safety: This is safe since the lifetime of the Future
+                    // is totally constraint to the lifetime of the select!
+                    // expression, and the Future can't get moved inside it
+                    // (it is shadowed).
                     future_let_bindings.push(quote! {
                         let mut #variant_name = #expr;
+                        let mut #variant_name = unsafe {
```

Comment:
Rather than creating an actual mutable reference here, which will cost extra stack space and indirection, what if we had the branch above (for direct bindings) add something like `future_let_bindings.push(quote! { #futures_crate::assert_type_behind_mut_ref_is_unpin(&mut path)})`. Then later we can use `Pin::new_unchecked` in each call to `poll`, since we already know each value being used must either be `Unpin` or be pinned.

Reply:
Interesting idea. But I'm not sure whether it's really worth it. The pin creation here is just a pointer, and I would assume the compiler to observe where this is really used and it can also just put it into registers. I won't expect this to use any extra stack. But maybe we can confirm it.

The main downside of this proposal for me would be that we split the spread out the `unsafe` code over more places: We would introduce a reliance of the `poll()` call on the setup having verified all preconditions.

Reply:
It will not do that-- since it's being held across an `await`, it will most definitely use not just stack space but increase the size of the generated future.

Reply:
I implemented this now in a branch at https://github.com/Matthias247/futures-rs/compare/fix_select_mutability...Matthias247:fix_select_mutability_2?expand=1

Is that what you had been thinking about? I can't see it making any difference in the generated future size. It is 16 bytes for both approaches in the test where I added an assertion. If that's true then it doesn't matter which way we go. Then just let me know if I should add the commit too, or if we merge as it is.

If I'm missing something let me know too.

Reply:
It isn't changing size there because there's a `default` branch, which means that `async` block contains no `await`s and its size is only determined by the pointers to the variables it captures + the generator state index. If you remove the default branch, the size is 48 with that change, and 64 without it.

Reply:
Oh ok. I unfortunately missed that part. I will add that commit then.

Reply:
Thanks! <3

Reply:
Thanks! <3

Reply:
updated